pkg/diag/validator/config_connector.go

```diff
@@ -60,58 +56,51 @@ func (ccv *ConfigConnectorValidator) Validate(ctx context.Context, ns string, op
-	status := &configConnectorResourceStatus{
-		name:      res.GetName(),
-		namespace: res.GetNamespace(),
-		ae: proto.ActionableErr{
-			ErrCode: proto.StatusCode_STATUSCHECK_SUCCESS,
-		},
-	}
-
+func getResourceStatus(res unstructured.Unstructured) (kstatus.Status, *proto.ActionableErr) {
 	// config connector resource statuses follow the Kubernetes kstatus so we use the attached kstatus library
 	// https://github.com/kubernetes-sigs/cli-utils/tree/master/pkg/kstatus#the-ready-condition
 	result, err := kstatus.Compute(&res)
 	if err != nil || result == nil {
-		status.result = kstatus.Result{Status: kstatus.UnknownStatus}
-		status.updateAE(proto.StatusCode_STATUSCHECK_UNKNOWN, "unable to check resource status")
-		return status
+		return kstatus.UnknownStatus, &proto.ActionableErr{ErrCode: proto.StatusCode_STATUSCHECK_UNKNOWN, Message: "unable to check resource status"}
 	}
-	status.result = *result
+	var ae proto.ActionableErr
 	switch result.Status {
 	case kstatus.CurrentStatus:
-		return status
+		ae = proto.ActionableErr{ErrCode: proto.StatusCode_STATUSCHECK_SUCCESS}
 	case kstatus.InProgressStatus:
 		if result.Message == "" {
-			status.updateAE(proto.StatusCode_STATUSCHECK_CONFIG_CONNECTOR_IN_PROGRESS, result.Message)
+			ae = proto.ActionableErr{ErrCode: proto.StatusCode_STATUSCHECK_CONFIG_CONNECTOR_IN_PROGRESS, Message: result.Message}
 		} else {
-			// config connector status doesn't always correctly parse to failed, but shows InProgress with an error message
-			status.updateAE(proto.StatusCode_STATUSCHECK_CONFIG_CONNECTOR_FAILED, result.Message)
+			// TODO: config connector resource status doesn't distinguish between resource that is making progress towards reconciling from one that is doomed.
+			// This is tracked in b/187759279 internally. As such to avoid stalling the status check phase until timeout in case of a failed resource,
+			// we report an error if there's any message reported without the status being success. This can cause skaffold to fail even when resources
+			// are rightly in an InProgress state, say while adding new nodes.
+			ae = proto.ActionableErr{ErrCode: proto.StatusCode_STATUSCHECK_CONFIG_CONNECTOR_FAILED, Message: result.Message}
 		}
 
 	case kstatus.FailedStatus:
-		status.updateAE(proto.StatusCode_STATUSCHECK_CONFIG_CONNECTOR_FAILED, result.Message)
+		ae = proto.ActionableErr{ErrCode: proto.StatusCode_STATUSCHECK_CONFIG_CONNECTOR_FAILED, Message: result.Message}
 	case kstatus.TerminatingStatus:
-		status.updateAE(proto.StatusCode_STATUSCHECK_CONFIG_CONNECTOR_TERMINATING, result.Message)
+		ae = proto.ActionableErr{ErrCode: proto.StatusCode_STATUSCHECK_CONFIG_CONNECTOR_TERMINATING, Message: result.Message}
 	case kstatus.NotFoundStatus:
-		status.updateAE(proto.StatusCode_STATUSCHECK_CONFIG_CONNECTOR_NOT_FOUND, result.Message)
+		ae = proto.ActionableErr{ErrCode: proto.StatusCode_STATUSCHECK_CONFIG_CONNECTOR_NOT_FOUND, Message: result.Message}
```

Comment:
I wonder how this would happen!

Reply:
I don't actually think that we'd run into this. But I added it for completion's sake.